athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,286 @@
+            String blobName = blob.getName();
+            String folderPath = blobName.startsWith(path)
+                    ? blobName.replace(path, "")
+                    : blobName;
+            // remove the front-slash, because, the expression generated without it
+            if (folderPath.startsWith("/")) {
+                folderPath = folderPath.substring(1);
+            }
+            LOGGER.info("Examining folder {}", folderPath);
+            List<AbstractMap.SimpleImmutableEntry<String, String>> partitions = PartitionUtil.getPartitionColumnData(table, folderPath);
+            if (!partitions.isEmpty() && checkPartitionWithConstrains(partitions, expressions)) {
+                partitionFolders.add(partitions);
+                LOGGER.info("Folder {} is selected", folderPath);
+            }
+        }
+        return partitionFolders;
+    }
+
+    /**
+     * Retrieves a first of file has non-zero size
+     *
+     * @param bucket Name of the bucket
+     * @param prefix Prefix (aka, folder in Storage service) of the bucket from where this method with retrieve files
+     * @return A single file name under the prefix
+     */
+    protected Optional<String> getStorageFiles(String bucket, String prefix)
+    {
+        LOGGER.info("Listing nested files for prefix {} under the bucket {}", prefix, bucket);
+        Page<Blob> blobPage = storage.list(bucket, prefix(prefix.substring(1)));
+        for (Blob blob : blobPage.iterateAll()) {
+            if (isBlobFile(blob)) {
+               return Optional.of(blob.getName());
+            }
+        }
+        return Optional.empty();
+    }
+
+    // helpers
+    /**
+     * check whether it is file, It may return folder also
+     */
+    private boolean isBlobFile(Blob blob)
+    {
+        return blob.getSize() > 0;
+    }
+
+    @VisibleForTesting
+    protected Schema getFileSchema(String bucketName, String path, FileFormat format, BufferAllocator allocator)
+    {
+        requireNonNull(bucketName, "bucketName was null");
+        requireNonNull(path, "fileName was null");
+        LOGGER.info("Retrieving field schema from file {}, under the bucket {}", path, bucketName);
+        String uri = createUri(bucketName, path);
+        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
+                NativeMemoryPool.getDefault(), format, uri);
+        // inspect schema
+        return factory.inspect();
+    }
+
+    private boolean checkPartitionWithConstrains(List<AbstractMap.SimpleImmutableEntry<String, String>> partitionList, List<EqualsExpression> expressions)
+    {
+        if (expressions.isEmpty()) {
+            return true;
+        }
+        for (AbstractMap.SimpleImmutableEntry<String, String> partition : partitionList) {
+            Optional<EqualsExpression> optionalExpression = expressions.stream()
+                    .filter(expr -> expr.columnName.equalsIgnoreCase(partition.getKey()))
+                    .findFirst();
```

Comment:
Instead of iterating through `expressions` for every partition in the partitionList, turn this into a `TreeMap`, either before you pass it into this function if it is going to be used this way multiple times, or just do it after line 214 before the for loop, like this:

```
var expressionMap = expressions.stream()
    .map(expr -> expr.columnName)
    .collect(Collectors.toCollection(() ->
        new java.util.TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER)));
```

Then line 219, can just be:
```
EqualsExpression expression = expressionMap.get(partition.getKey());
if (expression != null) {
   // ...
   // line 221 doesn't need to exist anymore.
}
```

Reply:
Incorporated the changes.